### PR TITLE
Match: Implement ruby-git (gem) based repository management

### PR DESCRIPTION
**Pull request by tobiasstrebitzer**
# Problem
- At the moment, all git- related functionality for synchronizing certificates inside the match gem is run via system calls (FastlaneCore::CommandExecutor).
- While that works, it might be hard to maintain: Parsing of git cli responses and error handling need to be maintained and it might be hard make (and keep) it reliable, especially if new functionality is required (e.g. branch management, reverting, handling merge conflicts, etc...).
# Solution
- This PR replaces all current git- command invocations with functionality provided by the ruby-git gem.
- [Ruby/Git](https://github.com/schacon/ruby-git) is a Ruby library that can be used to create, read and manipulate Git repositories by wrapping system calls to the git binary.
# Downsides / Limitations
- ~~There is currently a bug (or rather limitation) where the git.status command will return an exception if run against an empty git repository without previous commits: https://github.com/schacon/ruby-git/issues/275. This can be overcome by either (1) patching this limitation in the ruby-git gem, (2) making sure git status is only called after at lease one commit was made, or (3) catching and handling the exception thrown by git-ruby.~~ fixed in https://github.com/fastlane/fastlane/pull/4517/commits/1370b3050304b7f69ad811538f80238e8f36854a 
- The latest commit in ruby-git was made on Feb 26 2016 (although i personally found this gem very reliable in the past, also considering that core git functionality is unlikely to change).

### CODE REVIEW

**Comment by KrauseFx** [comment]
This is super exciting, I agree that we should try to not use system calls. I'm a little worried about adding another dependency, especially a git gem dependency. We've made bad experience with that in https://github.com/danger/danger (cc @orta). 

Maybe @orta can add a short comment on which gems to avoid?

**Comment by KrauseFx** [comment]
> The latest commit in ruby-git was made on Feb 26 2016 (although i personally found this gem very reliable in the past, also considering that core git functionality is unlikely to change).

That's only a few months ago, so perfectly fine

**Comment by tobiasstrebitzer** [comment]
I can see that each new dependency builds up the risk of incompatibility. I'd say it also depends on how strongly git functionality could / should be used within fastlane. If dealing with git is a very unique case (e.g. only for synchronizing certificates in match) then it's quite a big dependency to add.

TL;DR

To elaborate: In our projects we're using git for more use cases, specifically updating the app source via git when running fastlane, checking out different branches depending on lane configuration, etc... Another use case is branch management in match to overcome the limitation of a single developer team, which we use to create apps for multiple clients / developer accounts (see #4499).

**Comment by tobiasstrebitzer** [comment]
There is a PR in git-ruby that would fix the bug described above: https://github.com/schacon/ruby-git/pull/205/. It's quite old though and never made it to master since dec 2014.

**Comment by orta** [comment]
`git` was the gem I settled on for Danger too, was small, understandable and had no native extensions

**Comment by KrauseFx** [comment]
Thanks @orta. @tobiasstrebitzer is this ready for review?

File: match/lib/match/git_helper.rb [hunk 1]
```diff
@@ -48,25 +46,19 @@ def self.match_version(workspace)
...
 
     def self.commit_changes(path, message, git_url)
-      Dir.chdir(path) do
-        return if `git status`.include?("nothing to commit")
+      git = Git.open(path)
 
-        Encrypt.new.encrypt_repo(path: path, git_url: git_url)
-        File.write("match_version.txt", Match::VERSION) # unencrypted
+      # Avoid calling git.status if no branch exists
+      return unless git.current_branch.nil? or git.status.any?
```
Comment (tobiasstrebitzer) [inline match/lib/match/git_helper.rb]: I'm checking whether git has a current_branch registered or not. Will receive the name of the branch if a commit has been made before, or nil if it's a newly created branch. The reason i'm doing this is because calling git.status on a new branch (without any commits) throws an error (due to a bug in ruby-git).

Comment (tobiasstrebitzer) [inline match/lib/match/git_helper.rb]: I'm checking whether git has a current_branch registered or not. Will receive the name of the branch if a commit has been made before, or nil if it's a newly created branch. The reason i'm doing this is because calling git.status on a new branch (without any commits) throws an error (due to a bug in ruby-git).

**Comment by tobiasstrebitzer** [comment]
@KrauseFx, ready for review ;-)

**Comment by KrauseFx** [comment]
Hey @tobiasstrebitzer, thanks again for your great contribution 🚀 The code looks great, I'll do some local testing over the next days, before merging this. Stay tuned 👍

**Comment by KrauseFx** [comment]
In general I love the idea of switching to a gem to work with git in a cleaner way, definitely something we should take over to the complete fastlane tool suite when using git.

File: match/lib/match/git_helper.rb [hunk 2]
```diff
@@ -48,25 +46,19 @@ def self.match_version(workspace)
     end
 
     def self.commit_changes(path, message, git_url)
-      Dir.chdir(path) do
-        return if `git status`.include?("nothing to commit")
```
Comment (KrauseFx) [inline match/lib/match/git_helper.rb]: Any reason why we got rid of this line? Will it behave fine if there were no changes?

**Comment by KrauseFx** [comment]
Alright, did some further review, this looks great. Before we can merge this, would you mind also updating the `git_branch` part of this PR and resolve the remaining merge conflicts? 👍 
This looks great already 🚀

**Comment by tobiasstrebitzer** [comment]
@KrauseFx i've merged in the latest master branch changes, and adopted to code to work with git_branch. Please review

**Comment by KrauseFx** [comment]
Thanks for the fast action, there seems to be a syntax error

```
SyntaxError: /Users/distiller/fastlane/match/spec/git_helper_spec.rb:12: syntax error, unexpected tLABEL, expecting keyword_do or '{' or '('
      git.commit("initial commit",o allow_empty: true)
```

**Comment by tobiasstrebitzer** [comment]
Sorry for that... lesson learned: don't edit files within github :-p

screen shot 2016-05-31 at 18 51 01

File: match/lib/match/git_helper.rb [hunk 3]
```diff
@@ -51,25 +49,19 @@ def self.match_version(workspace)
...
-        commands << "git add -A"
-        commands << "git commit -m #{message.shellescape}"
-        commands << "git push origin #{branch.shellescape}"
+      Encrypt.new.encrypt_repo(path: path, git_url: git_url)
+      File.write(File.join(path, "match_version.txt"), Match::VERSION) # unencrypted
 
-        UI.message "Pushing changes to remote git repo..."
+      UI.message "Pushing changes to remote git repo..."
+      git.add(all: true)
+      git.commit("message")
```
Comment (mfurtak) [inline match/lib/match/git_helper.rb]: should this be `message` instead of `"message"`?

**Comment by mfurtak** [comment]
This change seems great, but obviously affects the way that `match` is implemented at its most basic level. How can we introduce these changes slowly, without risking breaking many people's setups?

**Comment by KrauseFx** [comment]
Short update: we're working on a system to introduce major changes like this without affecting all users, we'll update the PR when we're ready cc @ohwutup @mpirri 👍

**Comment by ohwutup** [comment]
Hey @tobiasstrebitzer! So we wrote a little bit of code to allow for explicit 'Feature Gates' to be enabled/disabled by environment variables https://github.com/fastlane/fastlane/pull/5097/files. You can also user some extract-method refactoring and replace implementation of methods at run-time. 

We would really love to merge this pull request, but as mentioned, we would love to validate these changes before risking breaking people's setups. Would you mind refactoring the code a bit to use this new functionality? We would be happy to help make this happen if you don't have the time to bite off the work in the near future! 🚀

**Comment by ohwutup** [comment]
Hey @tobiasstrebitzer ! Just wanted to check in to see if you had some time to put this functionality behind the feature gating mechanism we have in place now! We would love to get this merged in and tested in the wild! 🐙

**Comment by KrauseFx** [comment]
Hey @tobiasstrebitzer, looks like you've been busy, I'll close the PR for now, however feel free to submit a new one if you've had the time to update the changes 👍 Thanks for all your contributions
